hollow/src/main/java/com/netflix/hollow/api/producer/metrics/AbstractProducerMetricsListener.java

```diff
@@ -0,0 +1,120 @@
+     * In a multiple producer setting, leader election typically favors long-lived leaders to avoid producer runs from frequently requiring
+     * to reload the full state before publishing data. When a cycle is skipped because the producer wasn't primary, the current value of
+     * no. of consecutive failures and most recent cycle success time are retained, and no cycle status (success or fail) is reported.
+     * @param reason Reason why the run was skipped
+     */
+    @Override
+    public void onCycleSkip(CycleSkipReason reason) {
+        cycleMetricsBuilder
+                .setConsecutiveFailures(consecutiveFailures)
+                .setIsCycleSuccessOptional(Optional.empty())
+                .setCycleDurationMillisOptional(OptionalLong.empty())
+                .setLastCycleSuccessTimeNanoOptional(lastCycleSuccessTimeNanoOptional);
+
+        cycleMetricsReporting(cycleMetricsBuilder.build());
+    }
+
+    /**
+     * Reports announcement-related metrics.
+     * @param status Indicates whether the announcement succeeded of failed
+     * @param readState Hollow data state that is being published, used in this method for computing data size
+     * @param version Version of data that was announced
+     * @param elapsed Announcement start to end duration
+     */
+    @Override
+    public void onAnnouncementComplete(com.netflix.hollow.api.producer.Status status, HollowProducer.ReadState readState, long version, Duration elapsed) {
+        boolean isAnnouncementSuccess = false;
+        long dataSizeBytes = 0l;
+
+        if (status.getType() == com.netflix.hollow.api.producer.Status.StatusType.SUCCESS) {
+            isAnnouncementSuccess = true;
+            lastAnnouncementSuccessTimeNanoOptional = OptionalLong.of(System.nanoTime());
+        }
+
+        HollowReadStateEngine stateEngine = readState.getStateEngine();
+        dataSizeBytes = stateEngine.calcApproxDataSize();
+
+        announcementMetricsBuilder
+                .setDataSizeBytes(dataSizeBytes)
+                .setIsAnnouncementSuccess(isAnnouncementSuccess)
+                .setAnnouncementDurationMillis(elapsed.toMillis())
+                .setLastAnnouncementSuccessTimeNanoOptional(lastAnnouncementSuccessTimeNanoOptional);
+
+        announcementMetricsReporting(announcementMetricsBuilder.build());
+    }
+
+    /**
+     * On cycle completion this method reports cycle metrics.
+     * @param status Whether the cycle succeeded or failed
+     * @param readState Hollow data state published by cycle, not used here because data size is known from when announcement completed
+     * @param version Version of data that was published in this cycle
+     * @param elapsed Cycle start to end duration
+     */
+    @Override
+    public void onCycleComplete(com.netflix.hollow.api.producer.Status status, HollowProducer.ReadState readState, long version, Duration elapsed) {
+        Optional<Boolean> isCycleSuccess;
```

Comment:
An empty value is never created. It's present with either true or false.

Reply:
Optional was convenient when the builder argument was Optional<Boolean>, but if that goes then this goes.